File: DSA/Tree/Code/PostorderTraversal.cpp
#include <iostream>
#include <stack>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    
    Node(int val) : data(val), left(NULL), right(NULL) {}
};

class PostorderTraversal {
public:
    // Recursive postorder: Left-Right-Root
    static void postorderRecursive(Node* root) {
        if (root == NULL) return;
        
        postorderRecursive(root->left);
        postorderRecursive(root->right);
        cout << root->data << " ";
    }
    
    // Iterative postorder using two stacks
    static void postorderIterative(Node* root) {
        stack<Node*> st1, st2;
        st1.push(root);
        
        while (!st1.empty()) {
            Node* node = st1.top();
            st1.pop();
            st2.push(node);
            
            if (node->left != NULL) st1.push(node->left);
            if (node->right != NULL) st1.push(node->right);
        }
        
        while (!st2.empty()) {
            cout << st2.top()->data << " ";
            st2.pop();
        }
    }
};

int main() {
    cout << "=== Postorder Traversal (C++) ===\n\n";
    
    // Create tree
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);
    
    cout << "Tree structure:\n";
    cout << "         1\n";
    cout << "        / \\\n";
    cout << "       2   3\n";
    cout << "      / \\ / \\\n";
    cout << "     4  5 6  7\n\n";
    
    cout << "Recursive Postorder (Left-Right-Root): ";
    PostorderTraversal::postorderRecursive(root);
    cout << "\n";
    
    cout << "Iterative Postorder (two stacks):      ";
    PostorderTraversal::postorderIterative(root);
    cout << "\n\n";
    
    cout << "=== Postorder Characteristics ===\n";
    cout << "1. Left subtree → Right subtree → Root\n";
    cout << "2. Root is processed last\n";
    cout << "3. Useful for deleting tree\n";
    cout << "4. Useful for postfix expression evaluation\n\n";
    
    cout << "=== Complexity Analysis ===\n";
    cout << "Time Complexity:  O(n) - visit each node once\n";
    cout << "Space Complexity: O(h) - h = height (recursion/stack)\n";
    cout << "                   O(n) worst case (skewed tree)\n";
    
    return 0;
}

File: DSA/Tree/Code/PostorderTraversal.cs
// Postorder Traversal in C#

using System;
using System.Collections.Generic;

class Node {
    public int Data { get; set; }
    public Node Left { get; set; }
    public Node Right { get; set; }
    
    public Node(int data) {
        Data = data;
        Left = null;
        Right = null;
    }
}

class PostorderTraversal {
    
    public static void PostorderRecursive(Node root) {
        if (root == null) return;
        
        PostorderRecursive(root.Left);
        PostorderRecursive(root.Right);
        Console.Write(root.Data + " ");
    }
    
    public static void PostorderIterative(Node root) {
        Stack<Node> stack1 = new Stack<Node>();
        Stack<Node> stack2 = new Stack<Node>();
        stack1.Push(root);
        
        while (stack1.Count > 0) {
            Node node = stack1.Pop();
            stack2.Push(node);
            
            if (node.Left != null) stack1.Push(node.Left);
            if (node.Right != null) stack1.Push(node.Right);
        }
        
        while (stack2.Count > 0) {
            Console.Write(stack2.Pop().Data + " ");
        }
    }
    
    static void Main() {
        Console.WriteLine("=== Postorder Traversal (C#) ===\n");
        
        // Create tree
        Node root = new Node(1);
        root.Left = new Node(2);
        root.Right = new Node(3);
        root.Left.Left = new Node(4);
        root.Left.Right = new Node(5);
        root.Right.Left = new Node(6);
        root.Right.Right = new Node(7);
        
        Console.WriteLine("Tree structure:");
        Console.WriteLine("         1");
        Console.WriteLine("        / \\");
        Console.WriteLine("       2   3");
        Console.WriteLine("      / \\ / \\");
        Console.WriteLine("     4  5 6  7\n");
        
        Console.Write("Recursive Postorder (Left-Right-Root): ");
        PostorderRecursive(root);
        Console.WriteLine();
        
        Console.Write("Iterative Postorder (two stacks):      ");
        PostorderIterative(root);
        Console.WriteLine("\n");
        
        Console.WriteLine("=== Postorder Characteristics ===");
        Console.WriteLine("1. Left subtree → Right subtree → Root");
        Console.WriteLine("2. Root is processed last");
        Console.WriteLine("3. Useful for deleting tree");
        Console.WriteLine("4. Useful for postfix expression evaluation\n");
        
        Console.WriteLine("=== Complexity Analysis ===");
        Console.WriteLine("Time Complexity:  O(n) - visit each node once");
        Console.WriteLine("Space Complexity: O(h) - h = height (recursion/stack)");
        Console.WriteLine("                   O(n) worst case (skewed tree)");
    }
}
